google/cloud/aiplatform/base.py

```diff
@@ -16,13 +16,196 @@
 #
 
 import abc
-from typing import Optional
+from concurrent import futures
+import functools
+import inspect
+import threading
+from typing import Any, Callable, Dict, Optional, Sequence, Union
 
 from google.auth import credentials as auth_credentials
 from google.cloud.aiplatform import utils
 from google.cloud.aiplatform import initializer
 
 
+class FutureManager(metaclass=abc.ABCMeta):
+    """Tracks concurrent futures against this object."""
+
+    def __init__(self):
+        self.__latest_future_lock = threading.Lock()
+
+        # Always points to the latest future. All submitted futures will always
+        # form a dependency on the latest future.
+        self.__latest_future = None
+
+        # Caches Exception of any executed future. Once one exception occurs
+        # all additional futures should fail and any additional invocations will block.
+        self._exception = None
+
+    def _raise_future_exception(self):
+        """Raises exception if one of the object's futures has raised."""
+        with self.__latest_future_lock:
+            if self._exception:
+                raise self._exception
+
+    def _complete_future(self, future: futures.Future):
+        """Checks for exception of future and removes the pointer if it's still latest.
+
+        Args:
+            future (futures.Future): Required. A future to complete.
+        """
+
+        with self.__latest_future_lock:
+            try:
+                future.result()  # raises
+            except Exception as e:
+                self._exception = e
+
+            if self.__latest_future is future:
+                self.__latest_future = None
+
+    def _are_futures_done(self) -> bool:
+        """Helper method to check to all futures are complete.
+
+        Returns:
+            True if no latest future.
+        """
+        with self.__latest_future_lock:
+            return self.__latest_future is None
+
+    def wait(self):
+        """Helper method to that blocks until all futures are complete."""
+        future = self.__latest_future
+        if future:
+            futures.wait([future], return_when=futures.FIRST_EXCEPTION)
+
+        self._raise_future_exception()
+
+    @property
+    def _latest_future(self) -> Optional[futures.Future]:
+        """Get the latest future if it exists"""
+        with self.__latest_future_lock:
+            return self.__latest_future
```

Comment:
How does locking work here? Is it simply that an error would be raised if someone else holds the lock at the execution time of the return statement here?

Reply:
If someone else holds the lock it will block until that lock is released.